src/savegame_json.cpp

```diff
@@ -1529,21 +1529,62 @@ void item::io( Archive& archive )
 
     // Migrate legacy toolmod flags
     if( !is_toolmod() ) {
-        if( has_flag( "ATOMIC_AMMO" ) ) {
+        if( item_tags.count( "ATOMIC_AMMO" ) ) {
             item_tags.erase( "ATOMIC_AMMO" );
+            item_tags.erase( "NO_UNLOAD" );
+            item_tags.erase( "RADIOACTIVE" );
+            item_tags.erase( "LEAK_DAM" );
             emplace_back( "battery_atomic" );
 
-        } else if( has_flag( "DOUBLE_AMMO" ) ) {
+        } else if( item_tags.count( "DOUBLE_REACTOR" ) ) {
+            item_tags.erase( "DOUBLE_REACTOR" );
+            item_tags.erase( "DOUBLE_AMMO" );
+            emplace_back( "double_plutonium_core" );
+
+        } else if( item_tags.count( "DOUBLE_AMMO" ) ) {
             item_tags.erase( "DOUBLE_AMMO" );
             emplace_back( "battery_compartment" );
 
-        } else if( has_flag( "USE_UPS" ) ) {
+        } else if( item_tags.count( "USE_UPS" ) ) {
             item_tags.erase( "USE_UPS" );
+            item_tags.erase( "NO_RELOAD" );
+            item_tags.erase( "NO_UNLOAD" );
             emplace_back( "battery_ups" );
 
-        } else if( has_flag( "DOUBLE_REACTOR" ) ) {
-            item_tags.erase( "DOUBLE_REACTOR" );
-            emplace_back( "double_plutonium_core" );
+        }
+    }
+
+    // Fix fallout from #18797, which exponentially duplicates migrated toolmods
+    if( is_toolmod() ) {
+        // duplication would add an extra toolmod inside each toolmod on load;
+        // delete the nested copies
+        if( typeId() == "battery_atomic" || typeId() == "battery_compartment" ||
+            typeId() == "battery_ups" || typeId() == "double_plutonium_core" ) {
+            // Be conservative and only delete nested mods of the same type
+            contents.remove_if( [this]( const item &cont ) {
+                    return cont.typeId() == typeId();
+                        } );
+        }
+    }
+
+    if( is_tool() ) {
```

Comment:
Arguably you could remove everything except the first toolmod which would be simpler code. That said you could just refactor this entire migration code into a function and call it at which point the cautious approach doesn't harm readbility